google-cloud-pubsub/lib/google/cloud/pubsub/async_publisher.rb

```diff
@@ -224,40 +282,93 @@ def run_background
+        def resolve_batch_for_ordering_key ordering_key
+          @batches[ordering_key]
+        end
+
+        def publish_batches! stop: nil
+          @batches.reject! { |_ordering_key, batch| batch.empty? }
+          @batches.values.each do |batch|
+            ready = batch.publish! stop: stop
+            publish_batch_async @topic_name, batch if ready
+          end
+          # Set published_at to nil to wait indefinitely
+          @published_at = nil
         end
 
         def publish_batch_async topic_name, batch
+          # TODO: raise unless @publish_thread_pool.running?
           return unless @publish_thread_pool.running?
 
           Concurrent::Promises.future_on(
             @publish_thread_pool, topic_name, batch
-          ) do |t_name, btch|
-            publish_batch_sync t_name, btch
-          end
+          ) { |t, b| publish_batch_sync t, b }
         end
 
+        # rubocop:disable Metrics/AbcSize
+        # rubocop:disable Metrics/MethodLength
+
         def publish_batch_sync topic_name, batch
-          grpc = @service.publish topic_name, batch.messages
-          batch.items.zip Array(grpc.message_ids) do |item, id|
-            next unless item.callback
+          # The only batch methods that are safe to call from the loop are
+          # rebalance! and flush! because they are the only methods that are
+          # synchronized.
+          loop do
+            items = batch.rebalance!
+
+            unless items.empty?
+              grpc = @service.publish topic_name, items.map(&:msg)
+              items.zip Array(grpc.message_ids) do |item, id|
+                next unless item.callback
+
+                item.msg.message_id = id
+                publish_result = PublishResult.from_grpc item.msg
+                execute_callback_async item.callback, publish_result
+              end
+            end
 
-            item.msg.message_id = id
-            publish_result = PublishResult.from_grpc item.msg
-            execute_callback_async item.callback, publish_result
+            break unless batch.reset!
           end
         rescue StandardError => e
-          batch.items.each do |item|
+          items = batch.items
+
+          unless batch.ordering_key.empty?
+            retry if publish_batch_error_retryable? e
+            # Cancel the batch if the error is not to be retried.
+            begin
+              raise OrderingKeyError, batch.ordering_key
+            rescue OrderingKeyError => ordered_stopped
+              # Set the error to OrderingKeyError
+              e = ordered_stopped
+              # Get all unsent messages for the callback
+              items = batch.cancel!
+            end
+          end
+
+          items.each do |item|
             next unless item.callback
 
             publish_result = PublishResult.from_error item.msg, e
             execute_callback_async item.callback, publish_result
           end
+
+          retry unless batch.reset!
```

Comment:
I think you want `retry if batch.reset!`? You want to continue processing _if_ there is more to do?

Reply:
Yes, great catch. It should retry as long as there are items in the (reset) batch.